### PR TITLE
PARQUET-2369: Add new test file concatenated_gzip_members.parquet

### PR DESCRIPTION
**Pull request by amassalha**
For testing parquet file with multiple gzip members, following discussion in PARQUET-2369.

### CODE REVIEW

File: data/README.md [hunk 1]
```diff
@@ -45,6 +45,7 @@
 | plain-dict-uncompressed-checksum.parquet         | uncompressed and dictionary-encoded INT32 and STRING columns in format v1 with a matching CRC |
 | rle-dict-uncompressed-corrupt-checksum.parquet   | uncompressed and dictionary-encoded INT32 and STRING columns in format v2 with a mismatching CRC |
 | large_string_map.brotli.parquet       | MAP(STRING, INT32) with a string column chunk of more than 2GB. See [note](#large-string-map) below |
+| concatenated_gzip_members.parquet     | 513 int64 numbers compressed using 2 gzip members |
```
Comment (wgtmac) [inline data/README.md:48]:
```suggestion
| concatenated_gzip_members.parquet     | 513 UINT64 numbers compressed using 2 gzip members |
```

**Review by wgtmac — APPROVED** [review]
Thanks for adding this! I have left a minor comment with regard to the data type.

File: data/README.md (hunk 1 above)
Comment (amassalha) [reply data/README.md:48]: why you choosed [U]INT64 ?
parquet-mr shows INT64:

          type      encodings count     avg size   nulls   min / max
long_col  INT64     G RB_     513       2.86 B             "1" / "513"

Comment (wgtmac) [reply data/README.md:48]: The schema shows that its logical type is unsigned.
```
message root {
  optional int64 long_col (INTEGER(64,false));
}
```

Comment (amassalha) [reply data/README.md]: yes, I missed the logical type. thanks

**Review by wgtmac — APPROVED** [review]
LGTM +1.

cc @mapleFU @pitrou

**Comment by pitrou** [comment]
What does "2 gzip members" mean? The description is not really understandable as-is. You might expand in the description table, or create a separate paragraph to explain it in detail.

**Comment by amassalha** [comment]
@pitrou more details added

**Comment by pitrou** [comment]
Ok, but are these "gzip members" concatenated in a single data page? Which Parquet writer produces such files?

**Comment by pitrou** [comment]
I'll note that the [compression spec](https://github.com/apache/parquet-format/blob/master/Compression.md) doesn't say anything about potentially concatenating multiple GZip streams inside a single data page.

Given that data pages are typically small, there does not seem to be any reason for not compressing an entire data page at once.

@rdblue What do you think?

**Comment by amassalha** [comment]
> I'll note that the [compression spec](https://github.com/apache/parquet-format/blob/master/Compression.md) doesn't say anything about potentially concatenating multiple GZip streams inside a single data page.
> 
> Given that data pages are typically small, there does not seem to be any reason for not compressing an entire data page at once.
> 
> @rdblue What do you think?

see this:
https://www.zlib.net/manual.html
" inflate() will not automatically decode concatenated gzip members. inflate() will return Z_STREAM_END at the end of the **gzip member**. The state would need to be reset to continue decoding a subsequent gzip member. This must be done if there is more data after a gzip member, in order for the decompression to be compliant with the gzip standard (**RFC 1952**)."

**Comment by mapleFU** [comment]
I and Gang try to parse the file using arrow-rs and parquet-mr ( See https://github.com/apache/arrow/pull/38272#issuecomment-1767659431 ). Seems parquet-mr can read file like this, and arrow-rs report the bad page size. But I'm not sure:

1. arrow-rs report the decompressed size is a bit larger than size in page header, I need to recheck the size in the file
2. Whether the standard allow file like this.

**Comment by pitrou** [comment]
> its not about parquet data pages, its about gzip compressed data, size doesn't matter, you can concatenate members of any sizes. and then split to parquet pages if needed

Well, for now, you cannot. You are asking to complicate implementations to support this use case, so it's reasonable to ask: are there concrete reasons to support it?

**Comment by pitrou** [comment]
> I and Gang try to parse the file using arrow-rs and parquet-mr

How was the file produced?

**Comment by amassalha** [comment]
> Ok, but are these "gzip members" concatenated in a single data page? Which Parquet writer produces such files?


to generate this file we use  out own writer implementation (which is fully adheres to the gzip RFC)


> > I and Gang try to parse the file using arrow-rs and parquet-mr
> 
> How was the file produced?

its produced using our internal high-performance gzip implementation which is fully adheres to the gzip RFC.

**Comment by amassalha** [comment]
> concrete reasons

gzip/zlib standard (+we produce such files, parquet-mr + spark-shell can read this file)

**Comment by amassalha** [comment]
> > its produced using our internal high-performance gzip implementation which is fully adheres to the gzip RFC.
> 
> Is it deliberate that you produce several gzip fragments? This will decrease compression efficiency btw (since compressed fragments are independent from each other).

yes we undrestand that, its part of our internal implementation.

**Comment by tustvold** [comment]
So the parquet specification states

> A codec based on the GZIP format (not the closely-related "zlib" or "deflate" formats) defined by [RFC 1952](https://tools.ietf.org/html/rfc1952). If any ambiguity arises when implementing this format, the implementation provided by the [zlib compression library](https://zlib.net/) is authoritative.

The zlib decompression method states

> Unlike the gunzip utility and gzread() (see below), inflate() will not automatically decode concatenated gzip members. inflate() will return Z_STREAM_END at the end of the gzip member. The state would need to be reset to continue decoding a subsequent gzip member.

Given the specification makes no explicit mention of members, and the zlib library explicitly does not support them "out of the box", I would be inclined to read the specification as disallowing multiple members. Certainly it would appear multiple implementations, arrow-cpp and arrow-rs, interpreted the specification this way and regardless of the original author's intent, we should be optimising for maximum compatibility. 

I think if we want to alter the specification to support this, it should go through the usual process for a specification modification, including a clear articulation of the benefit to the wider ecosystem to incentivize adoption.

**Comment by amassalha** [comment]
@tustvold regards your edit (" it seems very surprising that this would be used to compress multiple parts of a single parquet page..."):  unless you consider a parallel compression...

**Comment by pitrou** [comment]
@amassalha That's true, but given the average size of a Parquet data page, wouldn't it be more efficient to parallelize accross data pages or columns?

**Comment by amassalha** [comment]
> @amassalha That's true, but given the average size of a Parquet data page, wouldn't it be more efficient to parallelize accross data pages or columns?

Im not sure I can explain more in details (its related to the startup internal Arch), but we parallelize alot of the compression parts, even inside same page (and columns ofcurse).

btw, see this for parralel compression:
https://zlib.net/pigz/

**Comment by tustvold** [comment]
I don't think anyone disputes that being able to introduce more parallelism would be nice for some workloads, the question is whether this is permissible according to the standard. Unfortunately in places where the standard is vague, in the interests of interoperability we normally have to take a lowest common denominator approach, which in this case would be to not support this...

**Comment by pitrou** [comment]
I would also stress that it's probably beneficial for your implementation to be compatible with previous versions of existing readers @amassalha . So, unless it's really painful for you, you should probably avoid creating data pages with several concatenated GZip streams.

**Comment by pitrou** [comment]
@amassalha I think you're just overinterpreting the zlib docs and the Parquet specs. Saying "the GZIP codec is based on the GZIP format" does not imply "the GZIP codec should transparently decompress multiple members into a single data page".

**Comment by amassalha** [comment]
> @amassalha I think you're just overinterpreting the zlib docs and the Parquet specs. Saying "the GZIP codec is based on the GZIP format" does not imply "the GZIP codec should **transparently** decompress multiple members into a single data page".

see this :
" 2.2. File format

      A gzip file consists of a series of "**members**" (compressed data
      sets).  The format of each member is specified in the following
      section.  The members simply appear one after another in the file,
      with no additional information before, between, or after them."

its from [RFC152](https://datatracker.ietf.org/doc/html/rfc1952#page-5)
its explained very staritly without any "overinterpreting "

**Comment by tustvold** [comment]
I think this discussion has gotten a bit circular, as a path forward how about:

* We update the parquet specification to permit multiple members but explicitly call out that some readers may not support this, it is not recommended, and writers should provide a mechanism to disable this behaviour
* We update arrow-cpp and arrow-rs to support reading multiple members

**Comment by wgtmac** [comment]
> I think this discussion has gotten a bit circular, as a path forward how about:
> 
> * We update the parquet specification to permit multiple members but explicitly call out that some readers may not support this, it is not recommended, and writers should provide a mechanism to disable this behaviour
> * We update arrow-cpp and arrow-rs to support reading multiple members

+1. It would be safe to change the parquet specs explicitly before adding support.

**Comment by pitrou** [comment]
> * We update the parquet specification to permit multiple members but explicitly call out that some readers may not support this, it is not recommended, and writers should provide a mechanism to disable this behaviour

+1, but I feel like there should be a discussion on the ML first.

**Comment by amassalha** [comment]
1) how should it be forwarded to parquet-format discussion ? should I add pull request to this: https://github.com/apache/parquet-format/blob/master/Compression.md ?

2) I still can't understand why you don't want to "fully" support this, while it clearly part of RC1952 that arrow is based on. how you decide when to "deviate" from RC1952 ?
we generate smaller parquet files very fast to save memory footprint and storage bandwidth. in this case we do not have enough parallelism from the number of pages, so we need sub-page compression

this scenario is supported in parquet-mr, spark-shell and by the gzip spec. why not by arrow?

**Comment by tustvold** [comment]
> why you don't want to "fully" support this

Because there is a substantial install base that won't support it, and users should be alerted to this fact, and be given options other than forcing an upgrade. It takes a long time for updates to percolate through the ecosystem...

**Comment by tustvold** [comment]
As a further datapoint, duckdb has a somewhat peculiar behaviour when confronted with this file

```
>>> duckdb.query("select * from 'concatenated_gzip_members.parquet'")
┌────────────┐
│  long_col  │
│   uint64   │
├────────────┤
│          1 │
│          2 │
│          3 │
│          4 │
│          5 │
│          6 │
│          7 │
│          8 │
│          9 │
│         10 │
│          · │
│          · │
│          · │
│        504 │
│        505 │
│        506 │
│        507 │
│        508 │
│        509 │
│        510 │
│        511 │
│        512 │
│          0 │
├────────────┤
│  513 rows  │
│ (20 shown) │
└────────────┘
```

It reports the final row as 0, not 513

**Comment by amassalha** [comment]
> As a further datapoint, duckdb has a somewhat peculiar behaviour when confronted with this file
> 
> ```
> >>> duckdb.query("select * from 'concatenated_gzip_members.parquet'")
> ┌────────────┐
> │  long_col  │
> │   uint64   │
> ├────────────┤
> │          1 │
> │          2 │
> │          3 │
> │          4 │
> │          5 │
> │          6 │
> │          7 │
> │          8 │
> │          9 │
> │         10 │
> │          · │
> │          · │
> │          · │
> │        504 │
> │        505 │
> │        506 │
> │        507 │
> │        508 │
> │        509 │
> │        510 │
> │        511 │
> │        512 │
> │          0 │
> ├────────────┤
> │  513 rows  │
> │ (20 shown) │
> └────────────┘
> ```
> 
> It reports the final row as 0, not 513

its same behaviour that libarrow cpp has before my patch

**Comment by mapleFU** [comment]
@tustvold I think this is same as arrow-cpp logic. I guess it's because during decompressing page, the page-size haven't been checked.

**Comment by mapleFU** [comment]
@pitrou @wgtmac Can we push this forward to merge this? Since the https://github.com/apache/parquet-format/pull/218 is merged.

**Comment by wgtmac** [comment]
It seems that rebase is required.

**Comment by amassalha** [comment]
> It seems that rebase is required.

I just rebased it

**Review by wgtmac — APPROVED** [review]
Thanks for rebasing.

@pitrou Do you have any comment?

File: data/README.md [hunk 2]
```diff
@@ -47,6 +47,7 @@
 | large_string_map.brotli.parquet       | MAP(STRING, INT32) with a string column chunk of more than 2GB. See [note](#large-string-map) below |
 | float16_nonzeros_and_nans.parquet | Float16 (logical type) column with NaNs and nonzero finite min/max values |
 | float16_zeros_and_nans.parquet    | Float16 (logical type) column with NaNs and zeros as min/max values. . See [note](#float16-files) below |
+| concatenated_gzip_members.parquet     | 513 UINT64 numbers compressed using 2 concatenated gzip members (A GZIP member refers to a single compressed file or stream that has been compressed using the GZIP algorithm.) |
```
Comment (pitrou) [inline data/README.md:50]:
```suggestion
| concatenated_gzip_members.parquet     | 513 UINT64 numbers compressed using 2 concatenated gzip members in a single data page |
```

**Comment by amassalha** [comment]
@pitrou  done

**Review by pitrou — APPROVED** [review]
LGTM as well.

**Pull request merged**